pennylane/grouping/utils.py

```diff
@@ -0,0 +1,603 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Utility functions used in Pauli partitioning and measurement reduction schemes utilizing the
+symplectic vector space representation of Pauli words. For information on the symplectic binary
+representation of Pauli words and applications, see:
+arXiv:quant-ph/9705052
+arXiv:1701.08213
+arXiv:1907.09386
+"""
+
+from pennylane import PauliX, PauliY, PauliZ, Identity
+from pennylane.operation import Observable, Tensor
+from pennylane.wires import Wires
+import numpy as np
+
+
+def get_n_qubits(observables):
+    """Obtains the number of qubits needed to support the given list of observables.
+
+    **Usage example:**
+
+    >>> observables = [qml.PauliX(0) @ qml.PauliY(13),
+                       qml.PauliZ(4) @ qml.PauliX(5)@ qml.PauliX(7)]
+    >>> get_n_qubits(observables)
+    14
+
+    Args:
+        observables (list[Observable]): list of observables with wires attribute.
+
+    Returns:
+        int: number of qubits.
+
+    Raises:
+        TypeError: if any item in observables list does not have the wires attribute.
+
+    """
+
+    for c, obs in enumerate(observables):
+        if not hasattr(obs, "wires"):
+            raise TypeError("Instance of {} does not have wires attribute.".format(type(obs)))
+        if isinstance(obs, Tensor):
+            observables[c] = obs.prune()
```

Comment:
I suppose the `Tensor` instance is pruned for the same reason that `Identity` instances are ignored when defining the dimensionality of the binary vector representation that I've described below. A set of Pauli words {X1 Y2 I3 I4, Z1 I2 I3 I4, Y1 X2 I3 I4} needs 2 qubits to be implemented, rather than 4 or higher. But if trivially acted on qubits are made 'explicit' by inclusion of an `Identity` instance, and we wish for the explicit identity operations to be considered differently than implicit identity operations, `get_n_qubits` can be changed to reflect that.